executor/projection.go

```diff
@@ -0,0 +1,258 @@
+			chk:          e.children[0].newChunk(),
+			targetWorker: e.workers[i],
+		}
+		e.fetcher.output <- &projectionOutput{
+			chk:  e.newChunk(),
+			done: make(chan error, 1),
+		}
+	}
+
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *ProjectionExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+
+	if e.numWorkers <= 0 {
+		err := e.children[0].Next(ctx, e.childrenResults[0])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = e.evaluatorSuit.Run(e.ctx, e.childrenResults[0], chk)
+		return errors.Trace(err)
+	}
+
+	if !e.prepared {
+		e.prepare(ctx)
+		e.prepared = true
+	}
+
+	output, ok := <-e.output
+	if !ok {
+		return nil
+	}
+
+	err := <-output.done
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	chk.SwapColumns(output.chk)
+	e.fetcher.output <- output
+	return nil
+}
+
+func (e *ProjectionExec) prepare(ctx context.Context) {
+	go e.fetcher.run(ctx)
+
+	e.waitGroup.Add(len(e.workers))
+	for i := range e.workers {
+		go e.workers[i].run(ctx)
+	}
+
+	go e.waitAndFinish()
+}
+
+func (e *ProjectionExec) waitAndFinish() {
+	e.waitGroup.Wait()
+	close(e.output)
+}
+
+// Close implements the Executor Close interface.
+func (e *ProjectionExec) Close() error {
+	if e.numWorkers > 0 {
+		close(e.fetcher.input)
+		close(e.fetcher.output)
+	}
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+type projectionInputFetcher struct {
+	child        Executor
+	globalOutput chan<- *projectionOutput
+
+	input  chan *projectionInput
+	output chan *projectionOutput
+}
+
+func (w *projectionInputFetcher) run(ctx context.Context) {
+	defer func() {
+		for input := range w.input {
+			input.targetWorker.finish()
+		}
+	}()
+
+	for {
+		input, ok1 := <-w.input
+		if !ok1 {
+			return
+		}
+		targetWorker := input.targetWorker
+
+		output, ok2 := <-w.output
+		if !ok2 {
+			targetWorker.finish()
+			return
+		}
+
+		w.globalOutput <- output
+
+		err := w.child.Next(ctx, input.chk)
+		if err != nil || input.chk.NumRows() == 0 {
+			output.done <- errors.Trace(err)
+			targetWorker.finish()
+			return
+		}
+
+		targetWorker.input <- input
+		targetWorker.output <- output
+	}
+}
+
+type projectionWorker struct {
+	sctx          sessionctx.Context
+	evaluatorSuit *expression.EvaluatorSuit
+	globalWG      *sync.WaitGroup
+	inputGiveBack chan<- *projectionInput
+
+	// channel "input" and "output" is :
+	// 0. inited  by "ProjectionExec.Open"
+	// 1. written by "projectionInputFetcher.run"
+	// 2. read    by "projectionWorker.run"
+	// 3. closed  by "projectionWorker.finish"
+	input  chan *projectionInput
+	output chan *projectionOutput
+}
+
+func (w *projectionWorker) run(ctx context.Context) {
+	defer w.globalWG.Done()
+
+	for {
+		input, ok1 := <-w.input
+		if !ok1 {
+			return
+		}
+
+		output, ok2 := <-w.output
+		if !ok2 {
+			return
+		}
+
+		err := w.evaluatorSuit.Run(w.sctx, input.chk, output.chk)
+		output.done <- errors.Trace(err)
+
+		if err != nil {
+			return
```

Comment:
If we return here,
projectionWorker.output will not be closed?

Reply:
It's OK, we don't need to close it